### PR TITLE
Hai Reveal: stormhold bugfix: "accompany" npcs are allowed to have disabled fighters that are docked in carriers

### LINKED ISSUE #7662: Get "on visit" message for Hai Reveal [B02] Join the Fleet when all marauders are dead and all "npc accompany" fleets are alive.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hai Reveal [B02] Join the Fleet is sporadically unplayable. I don't get "on fail" message, but I do get "on visit." All NPC accompany fleets are still alive, and all marauders are dead. 

**EDIT**: *With the merge of #7866, this should be fixed. To fix a broken save, depart the planet, repair the disabled fighters, and land again. I'm leaving this PR open a little longer to see if people find the mission breaking in new ways.*

### Steps to Reproduce

1. Play mission "Hai Reveal [B02] Join the Fleet"
2. Kill all the marauders
3. Land.
4. "on visit" message appears since mission is not completed, but no "on fail" message

### Expected Behavior

Mission completes when all objectives are met:

1. Marauders are defeated
2. At least one ship from each "npc accompany" fleet survives
3. Visited the waypoint.
4. Deadline is not reached

### Screenshots

_No response_

### Link to save file

Times it did happen:
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048461/Friction.Diction.txt)
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048847/Friction.Diction.txt)
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048855/Friction.Diction.txt) - all NPC escorts survived
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048870/Friction.Diction.txt)

Times it did not happen:
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048848/Friction.Diction.txt)
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048851/Friction.Diction.txt)
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048869/Friction.Diction.txt) - all NPC escorts survived
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048871/Friction.Diction.txt)
- [Friction Diction.txt](https://github.com/endless-sky/endless-sky/files/10048872/Friction.Diction.txt)

These are save files just before the mission begins. Visit the spaceport to start the mission
- [Friction Diction~3026-02-20 TRY10 part 5 begin with armada.txt](https://github.com/endless-sky/endless-sky/files/10048874/Friction.Diction.3026-02-20.TRY10.part.5.begin.with.armada.txt)
- [Friction Diction~3026-02-19 TRY10 part 5 begin.txt](https://github.com/endless-sky/endless-sky/files/10048875/Friction.Diction.3026-02-19.TRY10.part.5.begin.txt)

### Operating System

```
Linux 5.0.0-37-generic #40~18.04.1-Ubuntu SMP Thu Nov 14 12:06:39 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

### Game Source

Built from source

### Game Version

a82899f0

### Additional Information

- This doesn't happen 100% of the time. See the save files for examples.  I'd guess it's about 50% of the time.
- Sometimes (but not always) landing again after the "on visit" message will cause the mission to succeed.
- In one of my tests, this happened even though no NPC escort ships died.
- Removing "accompany" from the "npc accompany" fleets does not solve the problem.
- Waiting for all escorts to reach the waypoint does not solve the problem.
- I have not checked all saves, but in at least one, the Marauder ships are still alive, despite negative hitpoints

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Bugfix:** Addresses #7662

This is one of three PRs that, together, should fix issue #7862

## Fix Details
People have reported that the stormhold battle in Hai Reveal often doesn't complete. In one instance, I found it was because a fighter was disabled, but docked in a carrier. Presently, the logic requires the fighter to not be disabled, so the mission didn't succeed. Of course, since the fighter is already docked, you can't board it to repair it.

The fix I came up with is to consider a docked ship to be accompanied, even if it is disabled. This makes sense because the ship will land inside its carrier, when the carrier lands.

This is co-authored by @Hurleveur 

~~See  #7848 for the other half of this fix: the carrier should repair the fighter, so it isn't disabled anymore. Unfortunately, that does not fix #7662 on its own, since the NPC actions for that fighter retain the `DISABLED` event forever.~~

EDIT: After further analysis, I'm pretty sure (93% certainty) that this will fix the Stormhold mission since there are no other situations where a fighter with a DISABLED event can break that particular mission. It won't fix other missions that get stuck in this corner case, but at least that mission will succeed.

## Testing Done
Tested a failed attempt at storming stormhold that should have succeeded.

I don't know if there are ways this fix could break things.

## Save File

[Friction Diction~3026-01-21  should work but does not.txt](https://github.com/endless-sky/endless-sky/files/10203381/Friction.Diction.3026-01-21.should.work.but.does.not.txt)

If #7852 is merged first, then that save file cannot be used to verify the fix since it prevents the fighters from having DISABLED in this specific situation.

## Performance Impact
One "if" with a bizarre xor became three "if" statements. May lose you a few nanoseconds.

### CODE REVIEW

**Comment by UnorderedSigh** [comment]
There is another half of this bug, for which I don't have a fix. If the fighter is in the carrier, it shouldn't be disabled since the carrier has already "boarded" it.

File: source/NPC.cpp [hunk 1]
```diff
@@ -500,12 +500,21 @@ bool NPC::HasSucceeded(const System *playerSystem, bool ignoreIfDespawnable) con
 					&& !(it->second & ShipEvent::ASSIST);
 			}
 			bool isHere = false;
+			bool isCarried = ship->CanBeCarried() && ship->GetParent();
```
Comment (Hurleveur) [inline source/NPC.cpp:503]: Shouldnt iscarried also check that the ship has no system, meaning that it is in facr in a carrier?
Right now this is more canBeCarried than actually isCarried, I think

**Review by Hurleveur — CHANGES_REQUESTED** [review]
We talked about it already, I think that this way makes a lot of sense.
We shouldnt use a xor anyway since that makes the code harder to read.

That said maybe the last two ifs could be grouped into one but its fine this way.

**Comment by UnorderedSigh** [comment]
> That said maybe the last two ifs could be grouped into one but its fine this way.

It is easier to mentally parse if they are split up since the two "else ifs" check different problems.

**Comment by UnorderedSigh** [comment]
Closing due to merge of #7866

**Pull request closed without merging**